ActiveStrageプラクティスの課題を提出します。

k-nakamura1984 opened the pull request:
下記を実装しました。
- ユーザー詳細、一覧でアップロードした画像を表示。
- 画像のリサイズ。
- ユーザー編集画面にアイコン画像を表示。
- 特定の拡張子以外のファイルをアップロードした際にエラーを表示。
- ファイルのアップロード容量を1MBまでに制限。

元のリモートリポジトリをフォークしたのが、2021年5月31日だったため、最新のフォーク元を取得して、課題提出用のブランチにマージしております。
Railsプラクティスで、他のブランチにgit checkoutできない | FJORD BOOT CAMP（フィヨルドブートキャンプ）

自分以外のコミットが含まれてしまっているのですが、対応として適切でしたでしょうか…？
お手数ですが、合わせてご確認をいただけますと幸いです。

JunichiIto committed:
OmniAuth関連のコードを削除
https://bootcamp.fjord.jp/announcements/465

JunichiIto committed:
git fetchは不要なのでREADMEから削除

JunichiIto committed:
Merge branch 'main' into 03-pagination

JunichiIto committed:
Merge branch '03-pagination' into 04-user_page

JunichiIto committed:
Merge branch '04-user_page' into 05-github_login

JunichiIto committed:
Merge branch '05-github_login' into 06-user_icon

JunichiIto committed:
Fix Layout/EmptyLineAfterMagicComment

k-nakamura1984 committed:
ActiveStorageのプラクティス、課題のために下記を実装しました。

ユーザー詳細、一覧でアップロードした画像を表示できるようにしました。

k-nakamura1984 committed:
ActiveStorageのプラクティスに際して、下記を実装しました。

画像のリサイズ。

k-nakamura1984 committed:
ActiveStorageプラクティスのため、下記を実装しました。

ユーザー編集画面にアイコン画像を表示。
特定の拡張子以外のファイルをアップロードした際にエラーを表示。
ファイルのアップロード容量を1MBまでに制限。

yoshitsugu commented on app/models/user.rb:
```diff
@@ -2,16 +2,19 @@
-    end
+  private
+
+  def icon_type
+    errors.add(:icon, :invalid_file_type) unless icon.blob.content_type.in?(%('image/jpg image/jpeg image/png image/gif'))
+  end
```
歓迎条件のチェックですね :+1: 
こちら、iconがないユーザーを作ろうとするとエラーになるようになってしまっているようです。
iconがある場合のみバリデーションがはしるようにしてもらえると助かります。

yoshitsugu commented on app/models/user.rb:
```diff
@@ -2,16 +2,19 @@
+    errors.add(:icon, :invalid_file_type) unless icon.blob.content_type.in?(%('image/jpg image/jpeg image/png image/gif'))
+  end
+
+  def icon_size
+    errors.add(:icon, :invalid_file_size) if icon.blob.byte_size > 1.megabytes
   end
```
こちらもiconがない場合も許容するようにおねがいします！

yoshitsugu commented on app/views/devise/registrations/_profile_fields.html.erb:
```diff
@@ -1,3 +1,10 @@
+<div class="field">
+  <%= f.label :icon %>
+  <%= image_tag @user.icon.variant(resize: "64x64"), class: ApplicationController if @user.icon.attached? unless @user.errors.any? %>
```
後置 `if` や `unless` をつなげることは思わぬバグの原因になりそうなので、やらないほうがいいかなーと思います！
論理式の `&&` や `||` でつなげるようにしましょう！

yoshitsugu commented on db/schema.rb:
```diff
@@ -2,15 +2,43 @@
-ActiveRecord::Schema.define(version: 2020_11_22_065204) do
+ActiveRecord::Schema.define(version: 2021_05_31_231050) do
+
+  create_table "active_storage_attachments", force: :cascade do |t|
+    t.string "name", null: false
+    t.string "record_type", null: false
+    t.integer "record_id", null: false
+    t.integer "blob_id", null: false
+    t.datetime "created_at", null: false
+    t.index ["blob_id"], name: "index_active_storage_attachments_on_blob_id"
+    t.index ["record_type", "record_id", "name", "blob_id"], name: "index_active_storage_attachments_uniqueness", unique: true
+  end
+
+  create_table "active_storage_blobs", force: :cascade do |t|
+    t.string "key", null: false
+    t.string "filename", null: false
+    t.string "content_type"
+    t.text "metadata"
+    t.integer "byte_size", null: false
+    t.string "checksum", null: false
+    t.datetime "created_at", null: false
+    t.string "service_name", null: false
+    t.index ["key"], name: "index_active_storage_blobs_on_key", unique: true
+  end
+
+  create_table "active_storage_variant_records", force: :cascade do |t|
+    t.integer "blob_id", null: false
+    t.string "variation_digest", null: false
+    t.index ["blob_id", "variation_digest"], name: "index_active_storage_variant_records_uniqueness", unique: true
+  end
```
この辺の、active_storageまわりのmigrationファイルがコミットにふくまれていないようで、 `db:migrate` してもテーブルが生成されないようです。確認してもらってもよいですか？

yoshitsugu reviewed (commented):
手元で確認しようとしたものの、うまく実行できませんでした。。
コメントしましたので、見直してみてください〜

k-nakamura1984 committed:
ActiveStorageプラクティスのため、下記を実装しました。

iconの登録がない場合は、バリデーションの対象外にするよう変更。
削除していたマイグレーションファイルをrails active_storage:installすることにより再生成し、migrationとschema.rbを整合。

k-nakamura1984 commented on db/migrate/20210719041541_create_active_storage_tables.active_storage.rb:
```diff
@@ -0,0 +1,39 @@
+# frozen_string_literal: true
+
+# This migration comes from active_storage (originally 20170806125915)
```
削除していたmigtationファイルをrails active_strage:installすることにより再生成しました。

k-nakamura1984 commented on db/schema.rb:
```diff
@@ -49,35 +49,6 @@
     t.string "picture"
   end
 
-  create_table "comments", force: :cascade do |t|
```
申し訳ありません。
なぜ下記箇所が削除されたか原因を把握できておりません。

変更して進めているRailsプラクティスのブランチからgit checkoutしたことに関連していると推測しております。

git checkoutすることでコミットした段階と同じファイル構成に変更されると認識しているのですが、これは誤りでしょうか。

yoshitsugu replied:
git checkout でブランチの最新コミットの状態になる、ということであれば問題ないと思います。
原因の分析には、 git blame(annotate) などでコミット履歴をたどってみるのもよいかと思います

k-nakamura1984 replied:
ありがとうございます。
git blameをしてコミット履歴を確認したところjunichiItoさんの変更でした。
mergeした最新のフォーク元が反映されていたようです。

また他にテストブランチを作成して、gitの挙動を確認しました。
別ブランチでディレクトリを新規作成した場合、他のブランチに移動してもディレクトリはそのまま表示されることに違和感というか慣れておらず、「本当にコミットした段階に戻っているのだろうか」と心配になっていたようです。

例）
my-follow_userブランチでrelationshipテーブルを作成する。各種ディレクトリやファイルが作成される。
my-follow_userブランチでcomit。my-user_iconブランチにchecuoutする。
my-user_iconブランチで作成してないrelationshipディレクトリが表示される。

「Gitはディレクトリではなくファイルの変更を追跡する」ということを念頭に、git checkoutの後はファイル構成の確認を行えたらと思います。

yoshitsugu commented on app/views/users/index.html.erb:
```diff
@@ -18,6 +19,7 @@
         <td><%= user.name %></td>
         <td><%= user.postal_code %></td>
         <td><%= user.address %></td>
+        <td><%= image_tag user.icon.variant(resize: "64x64"), class: ApplicationController if user.icon.attached? %></td>
```
ここで、 `active_storage_attachments` と `active_storage_blobs` がユーザーごとにSQLが発行される状態、いわゆるN+1問題が発生しそうです(まだ実際にためせてないです)。ActiveStorageのN+1問題対応も調べたらすぐでてくると思うので、トライしてみてください！

yoshitsugu reviewed (commented):
手元で試すのが明日になりそうなので、とりあえずコードで気になったことだけコメントしました！

k-nakamura1984 committed:
N+1問題の対策のため、with_attached_attachment_nameを追記しました

yoshitsugu reviewed (approved):
OKです :+1:
